Library/Homebrew/dependency.rb

```diff
@@ -112,6 +112,8 @@ def expand(dependent, deps = dependent.deps, cache_key: nil, ignore_missing: fal
 
       deps.each do |dep|
         next if dependent.name == dep.name
+        # avoid downloading build dependency bottles
+        next if dep.build? && dependent.pour_bottle? && Homebrew::EnvConfig.install_from_api?
```

Comment:
Seems odd to me that it only happens in the fetch stage and only for API installs. Something must be done differently somewhere for this inconsistency to exist.

Reply:
I think the correct fix should be createing the `Formula` object from json directly instead of from the downloaded bottle file IIUC, but that needs more work as the formula json API doesn't have any Linux dependencies.

Reply:
Ah so #12936 basically.

Reply:
Yeah, exactly. Should I remove the first commit or close this pull request?

Reply:
I'm fine with at least the second commit as a short-term fix - it's consistent to what we have elsewhere.

I wouldn't object to the first commit - but I feel like there's something wrong elsewhere if it doesn't happen at install time and only at fetch time.

Reply:
`dependent.deps` at here contains all the dependencies including build dependencies, this happens also when core tap is cloned
https://github.com/Homebrew/brew/blob/b533b380e5e48a2092a93fa843f4bfb4eda8ee20/Library/Homebrew/dependency.rb#L101

then at this line `to_formula` will cause an error
https://github.com/Homebrew/brew/blob/b533b380e5e48a2092a93fa843f4bfb4eda8ee20/Library/Homebrew/dependency.rb#L128

then start downloading bottle and retry
https://github.com/Homebrew/brew/blob/b533b380e5e48a2092a93fa843f4bfb4eda8ee20/Library/Homebrew/formula_installer.rb#L221-L228

The draft pr uses the huge json file(https://formulae.brew.sh/api/formula.json) but it lacks dependencies and the json file of each formula lacks dependencies for Linux. In order to create `Formula` from json, json files in formulae.brew.sh need more data.

But why not download the zip file of homebrew-core and use that as if the repo is cloned? It's about the same size of `formula.json`.

Reply:
Ok that makes sense. I'm fine with the PR then as a fix until we can get rid of the API fetch layer everywhere.